assembly/src/parsers/crypto_ops.rs

```diff
@@ -200,149 +195,56 @@ pub(super) fn parse_mtree_cwm(
+    // no longer needed => [R_new, R, V_new, ...]
+    span_ops.push(Operation::MovUp4);
+    span_ops.push(Operation::Drop);
+    span_ops.push(Operation::MovUp4);
+    span_ops.push(Operation::Drop);
 
-    // move the new value to the top of the stack => [V_new, R_new, R, ...]
+    // Move the new value to the top => [R_new, V_new, R ...]
+    span_ops.push(Operation::SwapW);
     span_ops.push(Operation::SwapW2);
+    span_ops.push(Operation::SwapW);
 
     Ok(())
 }
 
-/// Validates that two 4 word Merkle roots at the top of the stack are equal, then drops the
-/// duplicate. The stack is expected to be arranged as follows (from the top):
-/// - root of a Merkle tree, 4 elements
-/// - root of a Merkle tree, 4 elements
-///
-/// This operation takes 20 VM cycles.
-fn validate_and_drop_root(span_ops: &mut Vec<Operation>) {
-    // verify the provided root and the computed root are equal
-    span_ops.push(Operation::Dup7);
-    span_ops.push(Operation::Dup4);
-    span_ops.push(Operation::Eq);
-
-    span_ops.push(Operation::Dup7);
-    span_ops.push(Operation::Dup4);
-    span_ops.push(Operation::Eq);
-    span_ops.push(Operation::And);
-
-    span_ops.push(Operation::Dup6);
-    span_ops.push(Operation::Dup3);
-    span_ops.push(Operation::Eq);
-    span_ops.push(Operation::And);
-
-    span_ops.push(Operation::Dup5);
-    span_ops.push(Operation::Dup2);
-    span_ops.push(Operation::Eq);
-    span_ops.push(Operation::And);
-    span_ops.push(Operation::Assert);
-
-    // drop one of the duplicate roots
-    span_ops.push_many(Operation::Drop, 4);
-}
-
-/// This is a helper function for assembly operations that update the Merkle tree. It preserves the
-/// new node value so it can be left on the stack at the end of the assembly sequence and prepares
-/// the stack with the elements and ordering expected by the VM's MRUPDATE operation. The stack is
-/// expected to be arranged as follows (from the top):
+/// This is a helper function for assembly operations that update the Merkle tree. It prepares
+/// the stack with the elements expected by the VM's MRUPDATE operation. The stack is expected
+/// to be arranged as follows (from the top):
 /// - depth of the node, 1 element
 /// - index of the node, 1 element
-/// - new value of the node, 4 elements
 /// - root of the Merkle tree, 4 elements
+/// - new value of the node, 4 elements
 ///
 /// After the operations are executed, the stack will be arranged as follows:
+/// - old value of the node, 4 elements
 /// - depth of the node, 1 element
 /// - index of the node, 1 element
-/// - old value of the node, 4 elements
-/// - new value of the node, 4 elements
 /// - root of the Merkle tree, 4 elements
-/// - copy of the new value of the node, 4 elements
+/// - new value of the node, 4 elements
 ///
-/// This operation takes 22 VM cycles.
+/// This operation takes 4 VM cycles.
 fn prep_stack_for_mrupdate(span_ops: &mut Vec<Operation>, decorators: &mut DecoratorList) {
```

Comment:
I think the meaning of this function changed enough that we can rename it to something like `read_mtree_node`. We can also use this function in `mtree_set` because it does the same thing as on lines 104 - 110.

Reply:
Agreed. I've updated this function so that `mtree_get` too can use it.